src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,468 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+            }
+            last_visible_indexes[i] = consensus->last_visible_index();
+        }
+
+        // A read_result needs to be returned for every partition. Hence,
+        // the function can't return before calling
+        // `fetch_ntps_in_parallel`.
+        std::vector<read_result> results = co_await fetch_ntps_in_parallel(
+          _ctx.mgr,
+          _ctx.srv.get_replica_selector(),
+          std::move(requests),
+          _ctx.foreign_read,
+          _ctx.deadline,
+          _ctx.bytes_left,
+          _ctx.srv.memory(),
+          _ctx.srv.memory_fetch_sem());
+
+        // If we weren't able to read the last_visible_index for a partition
+        // before calling `fetch_ntps_in_parallel` then we need to
+        // return with an error for that partition.
+        for (auto [i, partition] : errored_partitions) {
+            results[i] = read_result(error_code::not_leader_for_partition);
+            results[i].partition = partition;
+        }
+
+        for (const auto& r : results) {
+            total_size += r.data_size_bytes();
+            if (r.error != error_code::none) {
+                has_error = true;
+            }
+        }
+
+        co_return query_results{
+          .last_visible_indexes = std::move(last_visible_indexes),
+          .has_error = has_error,
+          .results = std::move(results),
+          .total_size = total_size,
+        };
+    }
+
+    std::optional<size_t> register_waiters(const auto& request_indexes) {
+        for (size_t i : request_indexes) {
+            auto part = _ctx.mgr.get(_ctx.requests[i].ktp());
+            // If the partition can't be found then it's since been moved
+            if (!part) {
+                return {i};
+            }
+
+            auto consensus = part->raft();
+            if (!consensus) {
+                return {i};
+            }
+
+            auto offset = model::next_offset(_last_visible_indexes[i]);
+            auto waiter = consensus->visible_offset_monitor()
+                            .wait(offset, model::no_timeout, _as)
+                            // All exceptions are ignored here as this is only
+                            // used to signal the worker that another attempt to
+                            // read the partition should be made.
+                            .handle_exception([](const std::exception_ptr&) {});
+
+            ssx::spawn_with_gate(
+              _waiter_gate, [this, w = std::move(waiter), i]() mutable {
+                  return w.finally([this, i] {
+                      _request_indexes.push_back(i);
+                      _has_completed_waiters.signal();
+                  });
+              });
+        }
+
+        return {};
+    }
+
+    ss::future<worker_result> do_run() {
+        bool first_run{true};
+        bool has_error{false};
+        // A map of indexes in `requests` to their corresponding index in
+        // `_ctx.requests`.
+        std::vector<size_t> requests_map;
```

Comment:
this vector is probably fine but the below one is probably just at the edge oversized alloc. `read_result` is 128 bytes so at high partition density and single requests for lots of partitions it might be at the edge.

It's an edge case though

Reply:
`requests_map` should always be fine since it's just an array of `size_t`. I'd expect this would never get too large since it's only used when waiting for partitions to have more data. And if tons of partitions were getting more data constantly I'd expect that this section of code would never be hit.

For `_ctx.requests` I think we'll probably hit a oversized alloc at some point since querying 2,000 partitions on a single shard may not be too impossible in the near future. Not sure if I should preemptively switch to a fragmented_vector though or wait until it happens. I'd need to switch more then just new code over as we store the requests and responses of a fetch in a `std::vector` currently.

Reply:
> I'd need to switch more then just new code over as we store the requests and responses of a fetch in a std::vector currently.

Maybe lets keep it for a separate commit then.